lib/Parse/ParseExpr.cpp

```diff
@@ -2399,6 +2399,41 @@ parseClosureSignatureIfPresent(SmallVectorImpl<CaptureListEntry> &captureList,
     }
   }
 
+  if (!params)
+    return invalid;
+
+  // If this was a closure declaration (maybe even trailing)
+  // tuple parameter destructuring is one of the common
+  // problems, and is misleading to users, so it's imperative
+  // to detect any tuple splat or destructuring as early as
+  // possible and give a proper fix-it. See SE-0110 for more details.
+  for (unsigned i = 0, e = params->size(); i != e; ++i) {
+    auto *param = params->get(i);
+    if (!param->isInvalid())
```

Comment:
Since there might be other ways a parameter could be invalid, it's probably best to check that it's actually missing a name and that the type is a TupleTypeRepr.